wakuv2/missing_messages.go

```diff
@@ -0,0 +1,272 @@
+package wakuv2
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"slices"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	"google.golang.org/protobuf/proto"
+
+	"go.uber.org/zap"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/store"
+
+	"github.com/status-im/status-go/logutils"
+	"github.com/status-im/status-go/wakuv2/common"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol/pb"
+)
+
+const maxContentTopicsPerRequest = 10
+const maxAttemptsToRetrieveHistory = 3
+const delay = 10 * time.Second
+
+type TopicInterest struct {
+	peerID        peer.ID
+	pubsubTopic   string
+	contentTopics []string
+	lastChecked   time.Time
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (p TopicInterest) Equals(other TopicInterest) bool {
+	if p.peerID != other.peerID {
+		return false
+	}
+
+	if p.pubsubTopic != other.pubsubTopic {
+		return false
+	}
+
+	slices.Sort(p.contentTopics)
+	slices.Sort(other.contentTopics)
+
+	for i, contentTopic := range p.contentTopics {
+		if contentTopic != other.contentTopics[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (w *Waku) SetTopicsToVerifyForMissingMessages(peerID peer.ID, pubsubTopic string, contentTopics []string) {
+	w.topicInterestMu.Lock()
+	defer w.topicInterestMu.Unlock()
+
+	ctx, cancel := context.WithCancel(w.ctx)
+	newMissingMessageRequest := TopicInterest{
+		peerID:        peerID,
+		pubsubTopic:   pubsubTopic,
+		contentTopics: contentTopics,
+		lastChecked:   w.timesource.Now().Add(delay),
+		ctx:           ctx,
+		cancel:        cancel,
+	}
+
+	currMessageVerificationRequest, ok := w.topicInterest[pubsubTopic]
+	if !ok || !currMessageVerificationRequest.Equals(newMissingMessageRequest) {
+		if ok {
+			// If there is an ongoing request, we cancel it before replacing it
+			// by the new list. This can be probably optimized further by tracking
+			// the last time a content topic was synced, but might not be necessary
+			// since cancelling an ongoing request would mean cancelling just a single
+			// page of results
+			currMessageVerificationRequest.cancel()
+		}
+
+		w.topicInterest[pubsubTopic] = newMissingMessageRequest
+	}
```

Comment:
minor readability improve suggestion
```suggestion
	if ok && currMessageVerificationRequest.Equals(newMissingMessageRequest) {
	    return
	}
    if ok {
	    // If there is an ongoing request, we cancel it before replacing it
	    // by the new list. This can be probably optimized further by tracking
	    // the last time a content topic was synced, but might not be necessary
	    // since cancelling an ongoing request would mean cancelling just a single
	    // page of results
	    currMessageVerificationRequest.cancel()
    }
    
    w.topicInterest[pubsubTopic] = newMissingMessageRequest
```